src/Microsoft.AspNetCore.Server.Testing/Deployers/NanoDeployer/NanoDeployer.cs

```diff
@@ -0,0 +1,251 @@
+{
+    public class NanoDeployer : ApplicationDeployer
+    {
+        /// <summary>
+        /// Example: If the share path is '\\foo\bar', then this returns the full path to the
+        /// deployed folder. Example: '\\foo\bar\048f6c99-de3e-488a-8020-f9eb277818d9'
+        /// </summary>
+        private string _deployedFolderPathInFileShare;
+        private readonly NanoDeploymentParameters _deploymentParameters;
+        private string _remotePSSessionHelperScript;
+
+        public NanoDeployer(DeploymentParameters deploymentParameters, ILogger logger)
+            : base(deploymentParameters, logger)
+        {
+            _deploymentParameters = DeploymentParameters as NanoDeploymentParameters;
+            if (_deploymentParameters == null)
+            {
+                throw new InvalidOperationException($"Expected the deployment parameters of type {nameof(NanoDeploymentParameters)}.");
+            }
+
+            if (_deploymentParameters.ServerType != ServerType.IIS
+                && _deploymentParameters.ServerType != ServerType.Kestrel
+                && _deploymentParameters.ServerType != ServerType.WebListener)
+            {
+                throw new InvalidOperationException($"Server type {_deploymentParameters.ServerType} is not supported for Nano." +
+                    " Supported server types are Kestrel, IIS and WebListener");
+            }
+
+            if (string.IsNullOrWhiteSpace(_deploymentParameters.ServerName))
+            {
+                throw new ArgumentException($"Invalid value for {nameof(NanoDeploymentParameters.ServerName)}");
+            }
+
+            if (string.IsNullOrWhiteSpace(_deploymentParameters.ServerAccountName))
+            {
+                throw new ArgumentException($"Invalid value for {nameof(NanoDeploymentParameters.ServerAccountName)}");
+            }
+
+            if (string.IsNullOrWhiteSpace(_deploymentParameters.ServerAccountPassword))
+            {
+                throw new ArgumentException($"Invalid value for {nameof(NanoDeploymentParameters.ServerAccountPassword)}");
+            }
+
+            if (string.IsNullOrWhiteSpace(_deploymentParameters.NanoServerFileShare))
+            {
+                throw new ArgumentException($"Invalid value for {nameof(NanoDeploymentParameters.NanoServerFileShare)}");
+            }
+
+            if (string.IsNullOrWhiteSpace(_deploymentParameters.NanoServerRelativeExecutablePath))
+            {
+                throw new ArgumentException($"Invalid value for {nameof(NanoDeploymentParameters.NanoServerRelativeExecutablePath)}");
+            }
+        }
+
+        public override DeploymentResult Deploy()
+        {
+            // Publish the app to a local temp folder on the machine where the test is running
+            DotnetPublish();
+
+            var folderId = Guid.NewGuid().ToString();
+            _deployedFolderPathInFileShare = Path.Combine(_deploymentParameters.NanoServerFileShare, folderId);
```

Comment:
If you call this method twice and then dispose the object, only one of the deployments will be cleaned up because you throw away the previous deployment's information. Either make it a noop if called twice or throw